src/main/java/roboy/util/TelegramCommunicationHandler.java

```diff
@@ -0,0 +1,339 @@
+import org.telegram.telegrambots.api.objects.Update;
+import org.telegram.telegrambots.api.objects.User;
+import org.telegram.telegrambots.bots.TelegramLongPollingBot;
+import org.telegram.telegrambots.exceptions.TelegramApiException;
+import roboy.io.TelegramInput;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.*;
+
+/** Singleton Class For Telegram Bot */
+public class TelegramCommunicationHandler extends TelegramLongPollingBot implements Timeout.TimeoutObserver{
+    private final static Logger logger = LogManager.getLogger();
+
+    private static final String tokensPath = "/Users/Apple/botboy/tokens.json";//place path to your token file here
+    private static final int TYPING_TIME_LIMIT = 3; //SECONDS
+    private static final int INPUT_TIME_LIMIT = 5; //SECONDS
+
+    // CHAT ID ----- ITS MESSAGE
+//    private volatile List<Pair<String, String>> pairs = new ArrayList<>();
+    private volatile List<HashMap<String, String>> pairs = new ArrayList<>();
+    private List<Timeout> telegramTimeouts; //Timeouts
+
+    private TelegramCommunicationHandler(){
+        super();
+        telegramTimeouts = new ArrayList<>();
+    }
+
+    // Instance for singleton
+    private static TelegramCommunicationHandler instance;
+
+    public static TelegramCommunicationHandler getInstance(){
+        if(instance == null){//for speed purposes
+            synchronized (TelegramCommunicationHandler.class){//in here: for threadsafety
+                //to prevent magic edgecases ruining our day
+                if(instance == null) instance = new TelegramCommunicationHandler();
+            }
+        }
+        return instance;
+    }
+
+    /**
+     * Waits until specified time passed after the last message w.r.t. given chatId.
+     * @param chatID unique identifier for a chat.
+     */
+    private void handleTimeout(String chatID){
+        int millisecond = 1000;
+
+        // if the list is empty create the first timeout and set it
+        if(telegramTimeouts.isEmpty()){
+            Timeout t  = new Timeout(millisecond * INPUT_TIME_LIMIT);
+            t.setUnique(chatID);
+
+            telegramTimeouts.add(t);
+        }
+
+        boolean hasStarted = false;
+        for(Timeout t: telegramTimeouts){
+            if(t.getUnique().equals(chatID)){
+                hasStarted = true;
+                t.start(this);
+            }
+        }
+
+        if(!hasStarted){
+            // there is no timeout for given chatID so start one
+            Timeout t  = new Timeout(millisecond * INPUT_TIME_LIMIT);
+            t.setUnique(chatID);
+            t.start(this);
+
+            telegramTimeouts.add(t);
+        }
+    }
+
+    /**
+     * Receives the updates from telegram's api and called by it.
+     * @param update consist of an update of a chat.
+     */
+    @Override
+    public void onUpdateReceived(Update update) {
+        if(update.hasMessage()){
```

Comment:
you can flip the if and check for "! update.hasMessage()" first
if this is the case, do the logging and call "return" inside the if

there is no need to indent a large block of code

Reply:
@polinominom